### ISSUE TITLE
No signing credential for algorithms  

### ISSUE THREAD

**Issue by DureSameen** [IdentityServer/IdentityServer4#4561]
I have registered a valid certificate in startup as

 ```
var certCollection = certStore.Certificates.Find(X509FindType.FindByThumbprint, certificateThumbprint, false);
            if (certCollection.Count == 0)
            {
                throw new Exception("CertificateNotFound");
            }

            var cert = certCollection[0];

            services.AddIdentityServer()
                .AddSigningCredential(new X509SigningCredentials(cert))
           //.AddValidationKey(cert)...
```

But I am still getting

System.InvalidOperationException: No signing credential for algorithms (System.Collections.Generic.List`1[System.String]) registered.
   at IdentityServer4.Services.DefaultKeyMaterialService.GetSigningCredentialsAsync(IEnumerable`1 allowedAlgorithms)
   at IdentityServer4.Services.DefaultTokenCreationService.CreateHeaderAsync(Token token)
   at IdentityServer4.Services.DefaultTokenCreationService.CreateTokenAsync(Token token)
   at IdentityServer4.Services.DefaultTokenService.CreateSecurityTokenAsync(Token token)
   at IdentityServer4.ResponseHandling.TokenResponseGenerator.CreateAccessTokenAsync(ValidatedTokenRequest request)
   at IdentityServer4.ResponseHandling.TokenResponseGenerator.ProcessTokenRequestAsync(TokenRequestValidationResult validationResult)
   at IdentityServer4.ResponseHandling.TokenResponseGenerator.ProcessAsync(TokenRequestValidationResult request)
   at IdentityServer4.Endpoints.TokenEndpoint.ProcessTokenRequestAsync(HttpContext context)
   at IdentityServer4.Endpoints.TokenEndpoint.ProcessAsync(HttpContext context)
   at IdentityServer4.Hosting.IdentityServerMiddleware.Invoke(HttpContext context, IEndpointRouter router, IUserSession session, IEventService events)
   at IdentityServer4.Hosting.IdentityServerMiddleware.Invoke(HttpContext context, IEndpointRouter router, IUserSession session, IEventService events)
   at IdentityServer4.Hosting.MutualTlsEndpointMiddleware.Invoke(HttpContext context, IAuthenticationSchemeProvider schemes)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Cors.Infrastructure.CorsMiddleware.<Invoke>g__InvokeCoreAwaited|15_0(HttpContext context, Task`1 policyTask)
   at IdentityServer4.Hosting.BaseUrlMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)

**Comment by leastprivilege** [comment]
Try using our overload that load the cert from the store. Does that make a difference?

**Comment by DureSameen** [comment]
Which overloads, can I get some link? .. 

bytheway this is Asp.net core 3.1 application and identityServerBuilder.AddDeveloperSigningCredential(); is not working on local host too.

**Comment by leastprivilege** [comment]
> Which overloads, can I get some link? ..

`AddSigningCredential`

> is not working 

can you be more specific?

**Comment by DureSameen** [comment]
@leastprivilege you said
Try using our overload that load the cert from the store. Does that make a difference?

I tried to find that overload that load the cert from the store. But I could not . Please give me a link or hint.

My problem is tthis. When I use AddDeveloperSigningCredential();  or load a certificate from store and then assign to .AddSigningCredential(new X509SigningCredentials(cert)) both method does not work and give me same exception.


`System.InvalidOperationException: No signing credential for algorithms (System.Collections.Generic.List1[System.String]) registered. at IdentityServer4.Services.DefaultKeyMaterialService.GetSigningCredentialsAsync(IEnumerable1`

**Comment by leastprivilege** [comment]
Please try our sample host to repro the problem.

**Comment by DureSameen** [comment]
Hi..

I was using .net core 3.1 so I downgraded all packages to 3.1 as following.

```
<PackageReference Include="IdentityServer4" Version="3.1.0" />
    <PackageReference Include="IdentityServer4.AspNetIdentity" Version="3.1.0" />
    <PackageReference Include="IdentityServer4.EntityFramework" Version="3.1.0" />
    <PackageReference Include="Microsoft.AspNetCore.Identity.EntityFrameworkCore" Version="3.1.5" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="3.1.5" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Tools" Version="3.1.5">
```
I also recreated configuration database 

After that error vanished..

IdentityServer4 Version 4.0.0 is not working properly with .net core 3.1.

**Comment by leastprivilege** [comment]
Well - as I said - it works fine here in that repo. That's why I asked you to try to reproduce the problem with our host.

**Comment by WisdomCorp** [comment]
I am facing the same issue as the first post. I am on Mac os using Visual studio for mac.

**Comment by leastprivilege** [comment]
OK - unfortunately there is a bug in the exception message, so we cannot see what algorithm we are looking for. This is fixed and will be in 4.0.4.

For now - please check if both on the Client and API resource the AllowedSigningAlgorithm properties are empty or not.

**Comment by WisdomCorp** [comment]
Thank you Dominick, it started working after I made **AllowedSigningAlgorithm** null in Client Table.

**Comment by leastprivilege** [comment]
What was in there?

**Comment by WisdomCorp** [comment]
0 (zero) was there

**Comment by leastprivilege** [comment]
How did you create the schema? Using plain EF migrations - or something else.

Wondering how 0 got in there.

**Comment by WisdomCorp** [comment]
It was entered by me :(

**Long Story:** As system upgrade, and some upgrade issues, in order to debug, I recreated IdentityServer4 database table manually from files in this link [https://github.com/IdentityServer/IdentityServer4/tree/main/src/EntityFramework.Storage/migrations/SqlServer/Migrations](url)
While re-doing data setup it was surely put by me as I saw only "AllowIdentityToken..." not the full field name.

**Comment by SpeakOutLoud** [comment]
After debugging the identity server solution, I found out that in development mode, the identity server uses AddDeveloperSigningCredential by default, and all the clients should have the property AllowedIdentityTokenSigningAlgorithms = RSA256.

When setting the value then the error will vanish. 
Why would we set the type of signing algorithms to clients? In the previous version 3.1 this was set by the code by default.

Thank you!
